php/hireinsocial/src/HireInSocial/Infrastructure/Doctrine/DBAL/Application/Offer/DbalOfferQuery.php

```diff
@@ -38,16 +38,41 @@ public function findAll(OfferFilter $filter): Offers
             $queryBuilder->andWhere('s.slug = :specializationSlug');
         }
 
-        $offersData = $queryBuilder->orderBy('o.created_at', 'DESC')
+        if ($filter->remote()) {
+            $queryBuilder->andWhere('o.location_remote = true');
+        }
+
+        if ($filter->withSalary()) {
+            $queryBuilder->andWhere('o.salary IS NOT NULL');
+        }
+
+        $queryBuilder
             ->setMaxResults($filter->limit())
-            ->setFirstResult($filter->offset())
-            ->setParameters(
-                [
-                    'specializationSlug' => $filter->specialization(),
-                    'sinceDate' => $filter->sinceDate()->format('Y-m-d H:i:s'),
-                    'tillDate' => $filter->tillDate()->format('Y-m-d H:i:s'),
-                ]
-            )->execute()
+            ->setFirstResult($filter->offset());
+
+        if ($filter->isSorted()) {
+            foreach ($filter->sortByColumns() as $column) {
+                if ($column->is(OfferFilter::COLUMN_SALARY)) {
+                    $queryBuilder->addOrderBy('salary_max', $column->direction());
+                }
+
+                if ($column->is(OfferFilter::COLUMN_CREATED_AT)) {
+                    $queryBuilder->addOrderBy('o.created_at', $column->direction());
+                }
+            }
+        } else {
+            $queryBuilder->orderBy('o.created_at', 'DESC');
+        }
```

Comment:
I'm not sure what to do here, this method is coming more and more complex, it's definitely something that I don't want to break in the future but I'm not sure whats the best way to test it. 

a) write functional tests for page where this query is used (I don't like it, functional tests are expensive and they would need to cover at lest 4 different scenarios, each with few offers)
b) write integration tests only for this query (I would need to write at least 4 tests with few offers) 
c) write unit tests and mock connection to return query builder mock and check what methods are called when passing different filters. (I'm not sure if I like creating a mock that returns a mock) 

At this point option c  with one additional scenario from option a convinces me the most, but I don't think I've ever seen anyone mocking query builder and writing unit tests for db queries.

Reply:
Unit-testing a query/query-builder is very weak, and won't help you much further: consider going for approach **a** here, because this is an integration point with ORM/DB which cannot be further split into two units.